src/endpoint/s3/ops/s3_post_object_select.js

```diff
@@ -0,0 +1,206 @@
+    }
+
+    _transform(chunk, enc, cb) {
+        dbg.log2(this.header, chunk.toString());
+        this.push(chunk);
+        return cb();
+    }
+}
+
+/**
+ * https://docs.aws.amazon.com/AmazonS3/latest/API/API_SelectObjectContent.html
+ */
+// eslint-disable-next-line max-statements
+async function post_object_select(req, res) {
+
+    if (('S3Select' in nb_native()) === false) {
+        dbg.error("Request for object select but compiled without BUILD_S3SELECT=1");
+        res.statusCode = 501;
+        res.end();
+        return;
+    }
+
+    req.object_sdk.setup_abort_controller(req, res);
+    const agent_header = req.headers['user-agent'];
+    const noobaa_trigger_agent = agent_header && agent_header.includes('exec-env/NOOBAA_FUNCTION');
+    const encryption = s3_utils.parse_encryption(req);
+
+    const md_params = {
+        bucket: req.params.bucket,
+        key: req.params.key,
+        version_id: req.query.versionId,
+        encryption,
+    };
+    if (req.query.get_from_cache !== undefined) {
+        md_params.get_from_cache = true;
+    }
+    const object_md = await req.object_sdk.read_object_md(md_params);
+
+    const params = {
+        object_md,
+        obj_id: object_md.obj_id,
+        bucket: req.params.bucket,
+        key: req.params.key,
+        content_type: object_md.content_type,
+        noobaa_trigger_agent,
+        encryption,
+    };
+    if (md_params.get_from_cache) {
+        params.get_from_cache = true;
+    }
+
+    const read_stream = await req.object_sdk.read_object_stream(params, res);
+    if (read_stream) {
+        // if read_stream supports closing, then we handle abort cases such as http disconnection
+        // by calling the close method to stop it from buffering more data which will go to waste.
+        if (read_stream.close) {
+            req.object_sdk.add_abort_handler(() => read_stream.close());
+        }
+        read_stream.on('error', err => {
+            dbg.log0('read stream error:', err, req.path);
+            res.destroy(err);
+        });
+        const http_req_select_params = req.body.SelectObjectContentRequest;
+        const input_serialization = http_req_select_params.InputSerialization[0];
+        let input_format = null;
+        if ('CSV' in input_serialization) {
+            input_format = 'CSV';
+        } else if ('JSON' in input_serialization) {
+            input_format = 'JSON';
+        } else {
+            throw new S3Error(S3Error.InvalidArgument);
+        }
+        const select_args = {
+            query: http_req_select_params.Expression[0],
+            input_format: input_format,
+            input_serialization_format: http_req_select_params.InputSerialization[0][input_format][0]
+        };
+        const s3select = new S3SelectStream(select_args);
+        s3select.on('end', () => {
+            res.write(end_message);
+            res.end();
+        });
+        dbg.log3("select_args = ", select_args);
```

Comment:
Is there a logic why is it in debug level 3?
there is also a print in debug level 2
`dbg.log2("select res = "`

Reply:
There's no obvious guideline. I think this is "debuggy" enough to warrant level3.